packages/peregrine/lib/talons/CheckoutPage/PaymentInformation/useCreditCard.js

```diff
@@ -0,0 +1,440 @@
+            }
+        });
+    }, [formState.values, updateBillingAddress, cartId]);
+
+    const setPaymentDetailsInCache = useCallback(
+        paymentNonce => {
+            /**
+             * We dont save the nonce due to PII
+             */
+            const { details, description, type } = paymentNonce;
+            client.writeQuery({
+                query: getPaymentNonceQuery,
+                data: {
+                    cart: {
+                        __typename: 'Cart',
+                        id: cartId,
+                        paymentNonce: {
+                            details,
+                            description,
+                            type
+                        }
+                    }
+                }
+            });
+        },
+        [cartId, client, getPaymentNonceQuery]
+    );
+
+    const updateCCDetailsOnCart = useCallback(
+        braintreeNonce => {
+            const { nonce } = braintreeNonce;
+            updateCCDetails({
+                variables: {
+                    cartId,
+                    paymentMethod: 'braintree',
+                    paymentNonce: nonce
+                }
+            });
+        },
+        [updateCCDetails, cartId]
+    );
+
+    const onPaymentSuccess = useCallback(
+        nonce => {
+            setPaymentDetailsInCache(nonce);
+            /**
+             * Updating payment nonce and selected payment method on cart.
+             */
+            updateCCDetailsOnCart(nonce);
+            setStepNumber(5);
+        },
+        [setPaymentDetailsInCache, updateCCDetailsOnCart]
+    );
+
+    const onPaymentError = useCallback(
+        error => {
+            setStepNumber(0);
+            if (onError) {
+                onError(error);
+            }
+        },
+        [onError]
+    );
+
+    const onPaymentReady = useCallback(() => {
+        setDropinLoading(false);
+        setStepNumber(0);
+        if (onReady) {
+            onReady();
+        }
+    }, [onReady]);
+
+    /**
+     * Effects
+     */
+
+    /**
+     * Step 1 effect
+     *
+     * User has clicked the update button
+     */
+    useEffect(() => {
+        if (updateButtonClicked) {
+            if (isBillingAddressSame) {
+                setShippingAddressAsBillingAddress();
+            } else {
+                setBillingAddress();
+            }
+            setStepNumber(1);
+            setIsBillingAddressSameInCache();
+        }
+    }, [
+        updateButtonClicked,
+        isBillingAddressSame,
+        setShippingAddressAsBillingAddress,
+        setBillingAddress,
+        setIsBillingAddressSameInCache
+    ]);
+
+    /**
+     * Step 2 effect
+     *
+     * Billing address mutation has completed
+     */
+    useEffect(() => {
+        const billingAddressMutationCompleted =
+            billingAddressMutationCalled && !billingAddressMutationLoading;
+        if (billingAddressMutationCompleted && !billingAddressMutationErrors) {
```

Comment:
Can you simplify step 1 and step 2 into a single effect using async await? I get that there is some messiness occurring due to the fact that the dropin is a child yet we want it to invoke a callback (ie `shouldRequestPaymentNonce` but I think we can simplify some of these effects a little.

Reply:
I feel is verbose enough to understand what is happening. Having them separated into diff code blocks makes it easy to maintain/modify if needed.

Reply:
Sure no problem.